File: app/controllers/iop/transfer/impl/OrganisationController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import javax.inject.Inject;
import miscellaneous.cache.FacilityCache;
import miscellaneous.config.ConfigReader;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import system.interceptors.SensitiveDataPolicy;

public class OrganisationController extends BaseController {

    @Inject
    private WSClient wsClient;

    @Inject
    private ConfigReader configReader;

    @Inject
    private FacilityCache facilityCache;

    private URL parseUrl() throws MalformedURLException {
        return URI.create(configReader.getIopHost() + "/api/organisations?withFacilities=true").toURL();
    }

    @Restrict({ @Group("STUDENT"), @Group("TEACHER"), @Group("ADMIN") })
    @SensitiveDataPolicy(sensitiveFieldNames = { "internalPassword", "externalPassword" })
    public CompletionStage<Result> listOrganisations() throws MalformedURLException {
        URL url = parseUrl();
        WSRequest request = wsClient.url(url.toString());
        String localRef = configReader.getHomeOrganisationRef();

        Function<WSResponse, Result> onSuccess = response -> {
            JsonNode root = response.asJson();
            if (response.getStatus() != Http.Status.OK) {
                return internalServerError(root.get("message").asText("Connection refused"));
            }
            if (root instanceof ArrayNode node) {
                for (JsonNode org : node) {
                    ((ObjectNode) org).put("homeOrg", org.get("_id").asText().equals(localRef));

                    // Cache facility data for external organizations and set the password requirement flag
                    if (org.has("facilities") && org.get("facilities").isArray()) {
                        ArrayNode facilities = (ArrayNode) org.get("facilities");
                        for (JsonNode facility : facilities) {
                            ObjectNode facilityObj = (ObjectNode) facility;
                            JsonNode passwordNode = facility.get("externalPassword");
                            boolean hasPassword =
                                passwordNode != null &&
                                !passwordNode.isNull() &&
                                !passwordNode.asText().trim().isEmpty();

                            // Set the flag for a client to know if a password is required
                            facilityObj.put("externalPasswordRequired", hasPassword);

                            if (facility.has("_id") && hasPassword) {
                                String facilityId = facility.get("_id").asText();
                                String password = facility.get("externalPassword").asText();
                                facilityCache.storeFacilityPassword(facilityId, password);
                            }
                        }
                    }
                }
            }
            return ok(root);
        };
        return request.get().thenApplyAsync(onSuccess);
    }
}

File: app/controllers/iop/transfer/impl/OrganisationController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl

import miscellaneous.cache.FacilityCache
import miscellaneous.config.ConfigReader
import models.user.Role
import play.api.libs.json.*
import play.api.libs.ws.WSClient
import play.api.mvc.{Action, AnyContent, BaseController, ControllerComponents}
import security.scala.Auth.{AuthenticatedAction, authorized}
import system.interceptors.scala.SensitiveDataFilter

import java.net.URI
import javax.inject.Inject
import scala.concurrent.ExecutionContext

class OrganisationController @Inject() (
    wsClient: WSClient,
    configReader: ConfigReader,
    facilityCache: FacilityCache,
    authenticated: AuthenticatedAction,
    sensitiveDataFilter: SensitiveDataFilter,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController:

  private def parseUrl(): String =
    URI.create(s"${configReader.getIopHost}/api/organisations?withFacilities=true").toString

  def listOrganisations: Action[AnyContent] =
    authenticated
      .andThen(authorized(Seq(Role.Name.STUDENT, Role.Name.TEACHER, Role.Name.ADMIN)))
      .andThen(sensitiveDataFilter(Set("internalPassword", "externalPassword")))
      .async { _ =>
        val url      = parseUrl()
        val localRef = configReader.getHomeOrganisationRef

        wsClient.url(url).get().map { response =>
          if response.status != OK then
            val message = (response.json \ "message").asOpt[String].getOrElse("Connection refused")
            InternalServerError(message)
          else
            val root = response.json match
              case JsArray(orgs) =>
                JsArray(orgs.map { org =>
                  val orgObj = org.as[JsObject]
                  val isHomeOrg = (orgObj \ "_id").asOpt[String].contains(localRef)
                  val baseOrg = orgObj + ("homeOrg" -> JsBoolean(isHomeOrg))

                  // Cache facility data for external organizations and set the password requirement flag
                  val updatedOrg = (orgObj \ "facilities").asOpt[JsArray].fold(baseOrg) { facilities =>
                    val updatedFacilities = JsArray(facilities.value.map { facility =>
                      val facilityObj = facility.as[JsObject]
                      val externalPassword = (facilityObj \ "externalPassword").asOpt[String].filter(_.nonEmpty)
                      val hasPassword = externalPassword.isDefined

                      // Set the flag for a client to know if a password is required
                      val updatedFacility = facilityObj + ("externalPasswordRequired" -> JsBoolean(hasPassword))
                      // Cache the password if present
                      for
                        facilityId <- (facilityObj \ "_id").asOpt[String]
                         password   <- externalPassword
                      yield facilityCache.storeFacilityPassword(facilityId, password)
                      updatedFacility
                    })
                    baseOrg + ("facilities" -> updatedFacilities)
                  }

                  updatedOrg
                })
              case other => other

            Ok(root)
        }
      }

